torch/lib/THD/base/DataChannel.hpp

```diff
@@ -75,20 +75,52 @@ struct DataChannel {
 
   virtual bool init() = 0;
 
+  /**
+   * This is required for NCCL backend, since the destroy cannot be done before
+   * CUDA is unloaded since DataChannel is a static object.
+   */
+  virtual void destroy() = 0;
+
   virtual rank_type getRank() = 0;
   virtual rank_type getNumProcesses() = 0;
 
-  virtual void allGather(std::vector<at::Tensor>& output, at::Tensor& input,
+  // All gather multiple GPUs on a number of nodes
+  virtual void allGather(std::vector<at::Tensor>& input,
+                         std::vector<at::Tensor>& output,
+                         THDGroup groupId = THDGroupWORLD) = 0;
+  virtual void allGather(std::vector<at::Tensor>& output,
+                         at::Tensor& input,
+                         THDGroup group_id = THDGroupWORLD) = 0;
+  virtual void gather(std::vector<at::Tensor>& output,
+                      at::Tensor& input,
+                      rank_type dst_rank,
+                      THDGroup group_id = THDGroupWORLD) = 0;
+  virtual void scatter(std::vector<at::Tensor>& input,
+                       at::Tensor& output,
+                       rank_type src_rank,
+                       THDGroup group_id = THDGroupWORLD) = 0;
+  // All reduce multiple GPUs on a number of nodes
+  virtual void allReduce(std::vector<at::Tensor>& input,
+                         std::vector<at::Tensor>& output,
+                         THDReduceOp operation,
                          THDGroup group_id = THDGroupWORLD) = 0;
-  virtual void gather(std::vector<at::Tensor>& output, at::Tensor& input,
-                      rank_type dst_rank, THDGroup group_id = THDGroupWORLD) = 0;
-  virtual void scatter(std::vector<at::Tensor>& input, at::Tensor& output,
-                       rank_type src_rank, THDGroup group_id = THDGroupWORLD) = 0;
   virtual void allReduce(at::Tensor& data, THDReduceOp operation,
                          THDGroup group_id = THDGroupWORLD) = 0;
-  virtual void reduce(at::Tensor& data, THDReduceOp operation,
-                      rank_type dst_rank, THDGroup group_id = THDGroupWORLD) = 0;
-  virtual void broadcast(at::Tensor& data, rank_type src_rank,
+  // Reduce multiple GPUs on a number of nodes
+  virtual void reduce(std::vector<at::Tensor>& data,
```

Comment:
This doesn't specify which tensor in `dstRank` will hold the result. Is it the first one?